### ISSUE TITLE
Implement tests

### ISSUE THREAD

**Issue by jkhartshorne** [pushkin-consortium/pushkin-api#10]

**Comment by jkhartshorne** [comment]
@rwang20 I got an async error when I tried running the tests. 

> jest


>   ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
>     Attempted to log "Pushkin API listening on port 3000".
> 
> 
> 
>       at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:186:10)
>       at Server.<anonymous> (src/api.js:115:17)
> 
>  PASS  test/api.test.js (22.259s)
>   ControllerBuilder
>     ✓ test controller init (10ms)
>     ✓ test setPass, one adding to passAlongs (1ms)
>     ✓ test setDirectUse, one adding to directUses (1ms)
>     ✓ test setDefaultPasses, six addings to passAlongs
>     ✓ test useController (138ms)
>     ✓ test start (16ms)
> 
>   console.log src/api.js:47
>     API using controller
> 
>   console.log test/api.test.js:56
>     ready to access the http://localhost:3000/myexp/router1 to see if any response
> 
> Test Suites: 1 passed, 1 total
> Tests:       6 passed, 6 total
> Snapshots:   0 total
> Time:        28.156s
> Ran all test suites.
> Jest did not exit one second after the test run has completed.
> 
> This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
>

**Comment by jkhartshorne** [comment]
I'm still getting similar errors:

> jest
> 
> 
>   ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
>     Attempted to log "Pushkin API listening on port 3000".
> 
> 
> 
>       at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:186:10)
>       at Server.<anonymous> (src/api.js:115:17)
> 
>  PASS  test/api.test.js (6.929s)
>   ControllerBuilder
>     ✓ test controller init (4ms)
>     ✓ test setPass, one adding to passAlongs (1ms)
>     ✓ test setDirectUse, one adding to directUses
>     ✓ test setDefaultPasses, six addings to passAlongs (1ms)
>     ✓ test useController (74ms)
>     ✓ test start (6ms)
>     ✓ test the endpoint (21ms)
> 
>   console.log src/api.js:47
>     API using controller
> 
> Test Suites: 1 passed, 1 total
> Tests:       7 passed, 7 total
> Snapshots:   0 total
> Time:        8.975s
> Ran all test suites.
> 
>   ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
>     Attempted to log "API got request for [object Object]".
> 
>       46 | 			throw new Error('Unable to add controllers after the API has started.');
>       47 | 		console.log('API using controller');
>     > 48 | 		this.app.use(route, controller); }
>          | 		            ^
>       49 | 
>       50 | 	usePushkinController(route, pushkinController) {
>       51 | 		if (this.expressListening)
> 
>       at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:186:10)
>       at src/api.js:48:15
>       at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
>       at trim_prefix (node_modules/express/lib/router/index.js:317:13)
>       at node_modules/express/lib/router/index.js:284:7
>       at Function.process_params (node_modules/express/lib/router/index.js:335:12)
>       at next (node_modules/express/lib/router/index.js:275:10)
>       at _cookieSession (node_modules/cookie-session/index.js:129:5)
>       at Layer.handle [as handle_request] (node_modules/express/lib/router/layer.js:95:5)
>       at trim_prefix (node_modules/express/lib/router/index.js:317:13)
> 
> Jest did not exit one second after the test run has completed.
> 
> This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.

**Comment by jrwang20** [comment]
According to similar issue discussions, this is because the tests run and are marked as successful before the logs/assertions run. Thus the tests end and the logs don't have a chance to run.

Adding the async keyword and done() callback function may be the solution, but for now, it doesn't work.

**Comment by jkhartshorne** [comment]
The update helped, but I'm still getting an error:

> Test Suites: 1 passed, 1 total
> Tests:       7 passed, 7 total
> Snapshots:   0 total
> Time:        10.481s
> Ran all test suites.
>   console.log src/api.js:22
>     API got request for [object Object]
> 
> Jest did not exit one second after the test run has completed.
> 
> This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
>

**Comment by jrwang20** [comment]
![image](https://user-images.githubusercontent.com/47158965/77460860-16bbee80-6dd8-11ea-83b1-2821b514e4b1.png)

**Comment by jrwang20** [comment]
I thought it's not an error. Since we start the server and didn't close it, the Jest would not exit. In this circumstance, if we send the request to the server (like just visit http://localhost:3000/myexp), there will still be a response message shown in this console. But still, I will find some solutions to handle this.

**Comment by jkhartshorne** [comment]
Well, jest doesn't exit successfully. You are right, though. The problem is that api.start() opens an express server and never closes is. To close an express server, we need to [name the return value of app.listen()](https://stackoverflow.com/questions/14515954/how-to-properly-close-node-express-server). 

So now start() reads like this:

`
	start() {
		if (!this.initialized)
			throw new Error('The API hasn\'t been successfully initialized');
		this.expressListening = true;
		this.server = this.app.listen(this.expressPort, async () => {
			console.log(`Pushkin API listening on port ${this.expressPort}`);
		});
	}
`

Then we just have to run 

`            	api.server.close();
`

after the done() in the last test. 

However, this is a change to pushkin-api. I don't really know what the consequences will be.

**Comment by jrwang20** [comment]
I think the pushkin-api will work fine since actually we changed nothing. The app.listen() method have a return value "server". In most cases, we wouldn't use it and only run the method. Now we just got the return value and assigned it to the Pushkin-API object as an attribute. We didn't make any changes to the main codes and didn't use this server attribute to do anything but in the tests.
The only thing we need to take care of is that the "server" attribute should not conflict with others.
